### PR TITLE
[+] Primary Key - Support  Composite primary keys

### PR DESCRIPTION
**Pull request by lucasscariot**

### CODE REVIEW

**Comment by lucasscariot** [comment]
[(13) Users want to have composite PK key supported](https://trello.com/c/jk8Fsqf3/1487-13-users-want-to-have-composite-pk-key-supported)

File: services/resource-getter.js [hunk 1]
```diff
@@ -19,10 +23,20 @@ function ResourceGetter(model, params) {
...
+    var where = {};
+    if (schema.isCompositePrimary && params.recordId) {
+      var recordId = params.recordId.split('-');
+      if (recordId.length === _.keys(model.primaryKeys).length) {
+        _.keys(model.primaryKeys).forEach(function (key, index) {
+          where[key] = recordId[index];
+        });
+      } else { return; }
+    } else {
+      where.id = params.recordId;
```
Comment (arnaudbesnier) [inline services/resource-getter.js]: `id` is hardcoded while using `where.id`. Use `where[schema.idField]`

File: services/resource-getter.js [hunk 2]
```diff
@@ -19,10 +23,20 @@ function ResourceGetter(model, params) {
...
 
   this.perform = function () {
+    var where = {};
+    if (schema.isCompositePrimary && params.recordId) {
+      var recordId = params.recordId.split('-');
+      if (recordId.length === _.keys(model.primaryKeys).length) {
+        _.keys(model.primaryKeys).forEach(function (key, index) {
+          where[key] = recordId[index];
+        });
+      } else { return; }
```
Comment (arnaudbesnier) [inline services/resource-getter.js]: This edge case is not really handled. What happens if we enter in this case?

File: services/primary-composite-key.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+var _ = require('lodash');
+
+function PrimaryCompositeKeys(model, schema, record) {
```
Comment (arnaudbesnier) [inline services/primary-composite-key.js]: This is a `CompositeKeysManager` with which you can:

- generate the "fake" id of a record (the method `get` you defined but that is not really well-name IMO)
- extract the conditions to retrieve the record (the method you should have factorized here and that is duplicated in `resource-getter` and `resource-updater`)

File: services/primary-composite-key.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  var glue = '-';
+
+  this.get = function () {
+    var forestCompositePrimary = '';
+    _.keys(model.primaryKeys).forEach(function (key, index) {
```
Comment (arnaudbesnier) [inline services/primary-composite-key.js]: `key` should be renamed `primaryKey` to be more explicit.

File: services/primary-composite-key.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+var _ = require('lodash');
+
+function PrimaryCompositeKeys(model, schema, record) {
+  var glue = '-';
```
Comment (arnaudbesnier) [inline services/primary-composite-key.js]: If this is a constant, use capital letters.

File: services/resource-getter.js [hunk 6]
```diff
@@ -19,10 +23,20 @@ function ResourceGetter(model, params) {
...
+    } else {
+      where.id = params.recordId;
+    }
```
Comment (arnaudbesnier) [inline services/resource-getter.js]: TODO: Move this in your new service

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,9 @@
 # Change Log
 
 ## [Unreleased]
+### Added
+- Primary Composite Key - Primary Composite Key are supported by the liana.
```
Comment (arnaudbesnier) [inline CHANGELOG.md]: no need to specify `by the liana`, developers that will read the changelog know that it is related to the current liana.

File: services/resource-getter.js (hunk 1 above)
Comment (arnaudbesnier) [inline services/resource-getter.js]: According to wikipedia, the wording you can use is "composite key" or "composite primary key" (not "primary composite key").

Comment (arnaudbesnier) [inline services/resource-getter.js]: `Composite Keys - Support composite primary keys.`

File: services/resource-updater.js [hunk 8]
```diff
@@ -1,17 +1,30 @@
...
     var where = {};
-    where[schema.idField] = params[schema.idField];
+
+    if (schema.isCompositePrimary) {
+      _.keys(model.primaryKeys).forEach(function (key) {
+        where[key] = params[key];
+      });
+    } else {
+      where[schema.idField] = params[schema.idField];
+    }
```
Comment (arnaudbesnier) [inline services/resource-updater.js]: TODO: Move this in your new service

File: services/resource-getter.js [hunk 9]
```diff
@@ -19,19 +23,29 @@ function ResourceGetter(model, params) {
...
 
   this.perform = function () {
+    var where = {};
+
+    var KeysManager = new CompositeKeysManager(model, schema, params);
+    if (schema.isCompositePrimary && params.recordId) {
+      var PrimaryCompositeKeys = params.recordId.split('-');
+      if (PrimaryCompositeKeys.length === _.keys(model.primaryKeys).length) {
+        where = KeysManager.getCondition(PrimaryCompositeKeys, null);
+      }
```
Comment (arnaudbesnier) [inline services/resource-getter.js]: You can simplify more

**Pull request merged**